provider/pkg/schema/gen_tags.go

```diff
@@ -0,0 +1,73 @@
+package schema
+
+import (
+	"strings"
+
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+)
+
+type TagsStyle string
+
+const (
+	// TagsStyleUnknown indicates we can't identify the style of tags.
+	TagsStyleUnknown TagsStyle = ""
+	// TagsStyleUntyped is a style where the tags are represented as "Any" - without a schema.
+	TagsStyleUntyped TagsStyle = "untyped"
+	// TagsStyleStringMap is a style where the tags are represented as a map of strings.
+	TagsStyleStringMap TagsStyle = "stringMap"
+	// TagsStyleKeyValueArray is a style where the tags are represented as an array of key-value pairs.
+	TagsStyleKeyValueArray TagsStyle = "keyValueArray"
+)
+
+// GetTagsStyle identifies the style of tags based on the shape of the schema.
+func (ctx *context) GetTagsStyle(typeSpec *pschema.TypeSpec) TagsStyle {
+	if typeSpec == nil {
+		return TagsStyleUnknown
+	}
+	// Check for "Any" ref
+	if typeSpec.Ref == "pulumi.json#/Any" {
+		return TagsStyleUntyped
+	}
+
+	if typeSpec.AdditionalProperties != nil && typeSpec.AdditionalProperties.Type == "string" {
+		return TagsStyleStringMap
+	}
+
+	if ctx.tagStyleIsKeyValueArray(typeSpec) {
+		return TagsStyleKeyValueArray
+	}
+
+	return TagsStyleUnknown
+}
+
+func (ctx *context) tagStyleIsKeyValueArray(typeSpec *pschema.TypeSpec) bool {
+	if typeSpec == nil || typeSpec.Items == nil || typeSpec.Items.Ref == "" {
+		return false
+	}
+
+	typeToken, hasPrefix := strings.CutPrefix(typeSpec.Items.Ref, "#/types/")
+	if !hasPrefix {
+		return false
+	}
+
+	// We can't include tags which are create-only properties because this has to be added on the tags type but this is a shared type
+	if createOnlyProps := readPropSdkNames(ctx.resourceSpec, "createOnlyProperties"); createOnlyProps.Has("tags") {
+		return false
+	}
```

Comment:
I really don't like how this depends on the field name being "tags"

Reply:
I think this is safe enough given that we're only calling into this method when we're already looking at a property named "tags". We don't want to expand this to other property names because the type is specially named "Tag" so other types with the same fields wouldn't use this logic.

Reply:
Ok, fixed ... needed this now we changed the logic to also lookup alternative tags fields from the schema.